core/state_processor.go

```diff
@@ -98,23 +101,50 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 
+	// Overlay tree migration logic
+	migrdb := statedb.Database()
+	// TODO: avoid opening the preimages file here and make it part of, potentially, statedb.Database().
+	filePreimages, err := os.Open("preimages.bin")
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("opening preimage file: %s", err)
+	}
+	defer filePreimages.Close()
+	preimageSeek := migrdb.GetCurrentPreimageOffset()
+	if _, err := filePreimages.Seek(migrdb.GetCurrentPreimageOffset(), io.SeekStart); err != nil {
+		return nil, nil, 0, fmt.Errorf("seeking preimage file: %s", err)
+	}
+	fpreimages := bufio.NewReader(filePreimages)
+
 	// verkle transition: if the conversion process is in progress, move
 	// N values from the MPT into the verkle tree.
-	if statedb.Database().InTransition() {
+	if migrdb.InTransition() {
 		var (
 			now = time.Now()
 			tt  = statedb.GetTrie().(*trie.TransitionTrie)
 			mpt = tt.Base()
 			vkt = tt.Overlay()
 		)
 
-		accIt, err := statedb.Snaps().AccountIterator(mpt.Hash(), statedb.Database().GetCurrentAccountHash())
+		accIt, err := statedb.Snaps().AccountIterator(mpt.Hash(), migrdb.GetCurrentAccountHash())
 		if err != nil {
 			return nil, nil, 0, err
 		}
 		defer accIt.Release()
 		accIt.Next()
 
+		// If we're about to start with the migration process, we have to read the first account hash preimage.
+		if migrdb.GetCurrentAccountAddress() == (common.Address{}) {
+			var addr common.Address
+			if _, err := io.ReadFull(fpreimages, addr[:]); err != nil {
+				return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+			}
+			if crypto.Keccak256Hash(addr[:]) != accIt.Hash() {
```

Comment:
Not yet because `addr` is set as the current one in L145, so some lines before.

I could move L145 between L140 and L141 and then I could... if the early err-return in L142 happened, we'd have left this side effect which isn't "perfect".

But.. if that err-return happens, we'd have other huge problems... so I think I'll do it.

Reply:
Fixed in https://github.com/gballet/go-ethereum/pull/246/commits/c99f5355b2aa3eecc728a3680b8969c7441e8fb2.